hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,236 @@
+package is.hail.expr.ir.agg
+
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitTriplet, EmitRegion, EmitMethodBuilder}
+import is.hail.expr.types.physical._
+import is.hail.utils._
+
+object StagedBlockLinkedList {
+  val defaultBlockCap: Int = 64
+
+  // src : Code[PointerTo[T]]
+  // dst : Code[PointerTo[T]]
+  private def deepCopy(er: EmitRegion, typ: PType, src: Code[Long], dst: Code[Long]): Code[Unit] =
+    storeDeepCopy(er, typ, Region.loadIRIntermediate(typ)(src), dst)
+
+  // v : Code[T]
+  // dst : Code[PointerTo[T]]
+  private def storeDeepCopy(er: EmitRegion, typ: PType, v: Code[_], dst: Code[Long]): Code[Unit] = {
+    val ftyp = typ.fundamentalType
+    if (ftyp.isPrimitive)
+      Region.storePrimitive(ftyp, dst)(v)
+    else
+      StagedRegionValueBuilder.deepCopy(er, typ, coerce[Long](v), dst)
+  }
+}
+
+class StagedBlockLinkedList(val elemType: PType, val mb: EmitMethodBuilder) {
+  import StagedBlockLinkedList._
+
+
+  /*** State ***/
+
+  val firstNode = mb.newField[Long]
+  val lastNode = mb.newField[Long]
+  val totalCount = mb.newField[Int]
+
+  val storageType = PStruct(
+    "firstNode" -> PInt64Required,
+    "lastNode" -> PInt64Required,
+    "totalCount" -> PInt32Required)
+
+  def load(src: Code[Long]): Code[Unit] = Code(
+    firstNode := Region.loadAddress(storageType.fieldOffset(src, 0)),
+    lastNode := Region.loadAddress(storageType.fieldOffset(src, 1)),
+    totalCount := Region.loadInt(storageType.fieldOffset(src, 2)))
+
+  def store(dst: Code[Long]): Code[Unit] = Code(
+    Region.storeAddress(storageType.fieldOffset(dst, 0), firstNode),
+    Region.storeAddress(storageType.fieldOffset(dst, 1), lastNode),
+    Region.storeInt(storageType.fieldOffset(dst, 2), totalCount))
+
+
+  /*** Temporaries ***
+   * 'i': used to iterate over block buffers in foreachElemAddress
+   * 'p': used to track missingness in foreachBlockNode
+   * 'tmpNode': used to hold pointers to newly allocated nodes */
+  val i = mb.newField[Int]
+  val p = mb.newField[Boolean]
+  val tmpNode = mb.newField[Long]
+
+
+  /*** Node management ***/
+
+  type Node = Code[Long]
+
+  val bufferType = PArray(elemType.setRequired(false), required = true)
```

Comment:
I think you can keep the requiredness from the element type as long as you handle the missingness appropriately, right?